src/util/triplea/mapXmlCreator/CanalDefinitionsPanel.java

```diff
@@ -0,0 +1,450 @@
+package util.triplea.mapXmlCreator;
+
+import java.awt.Color;
+import java.awt.FontMetrics;
+import java.awt.Graphics;
+import java.awt.Graphics2D;
+import java.awt.Point;
+import java.awt.event.MouseEvent;
+import java.awt.geom.Rectangle2D;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Optional;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+import javax.swing.JOptionPane;
+import javax.swing.SwingUtilities;
+
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+
+import games.strategy.common.swing.SwingAction;
+import util.triplea.mapXmlCreator.TerritoryDefinitionDialog.DEFINITION;
+
+
+final public class CanalDefinitionsPanel extends ImageScrollPanePanel {
+  static final String HTML_CANAL_KEY_POSTFIX = ": ";
+  static final String HTML_CANAL_KEY_PREFIX = "<br/> - ";
+  private static final double PI_HALF = Math.PI / 2;
+  private static final String NEW_CANAL_OPTION = "<new Canal>";
+
+  private Set<String> selectedLandTerritories = new TreeSet<String>();
+  private Set<String> selectedWaterTerritories = new TreeSet<String>();
+  private Optional<String> currentCanalName = Optional.empty();
+
+  private CanalDefinitionsPanel() {}
+
+  public static void layout(final MapXmlCreator mapXMLCreator) {
+    ImageScrollPanePanel.mapXMLCreator = mapXMLCreator;
+    final CanalDefinitionsPanel panel = new CanalDefinitionsPanel();
+    panel.layout(mapXMLCreator.getStepActionPanel());
+    mapXMLCreator.setAutoFillAction(SwingAction.of(e -> {
+      final int prevCanalCount = MapXmlHelper.getCanalDefinitionsMap().size();
+      if (prevCanalCount > 0) {
+        if (JOptionPane.YES_OPTION != MapXmlUIHelper.showYesNoOptionDialog("Auto-Fill Warning",
```

Comment:
Favor using 'positive' conditions in conditions. Check for `== YES_OPTION` and do the clearCanalDefinitions if so, otherwise do somethign else. It's easier to understand conditions that are of type `if(true) { ... } else { }` rather than `if(!true) { ... } else { }`